## 모듈 import

In [1]:
from IPython.display import Image
import numpy as np
import pandas as pd
import seaborn as sns
from pandas import DataFrame

In [2]:
data = {
        'A' : ['foo', 'bar', 'foo', 'bar', 'foo', 'bar', 'foo', 'foo'],
        'B' : ['one', 'one', 'two', 'three', 'two', 'two', 'one', 'three'],
        'C' : [7, 2, 6, 0, 9, 9, 9, 7],
        'D' : [8, 1, 5, 9, 1, 5, 7, 3],
}
df = pd.DataFrame(data)

In [7]:
df.pivot_table(index='A',columns='B',aggfunc=['mean', 'sum'])

mean                            sum                        
       C               D              C             D          
B    one three  two  one three  two one three two one three two
A                                                              
bar  2.0   0.0  9.0  1.0   9.0  5.0   2     0   9   1     9   5
foo  8.0   7.0  7.5  7.5   3.0  3.0  16     7  15  15     3   6

In [44]:
df['C3'] = np.nan

In [45]:
df

C1  C2  C3        C4  (2, C2)
0 NaN NaN NaN -0.185253      NaN
1 NaN NaN NaN -0.281045      NaN
2 NaN NaN NaN  1.744754      NaN
3 NaN NaN NaN -0.431031      NaN
4 NaN NaN NaN -0.339469      NaN

In [36]:
df = DataFrame(np.random.randn(5, 4), columns=['C1', 'C2', 'C3', 'C4'])
df.dropna(axis=[0,1])

C1        C2        C3        C4
0  0.917195 -1.193398  0.289260 -0.185253
1 -0.769109 -0.863894  1.472465 -0.281045
2 -2.059731  1.062853  1.280529  1.744754
3  1.518326  1.337798  0.604524 -0.431031
4 -1.548204  2.350622 -0.784211 -0.339469

## 데이터셋 로드

In [2]:
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

**컬럼(columns) 설명**

- survivied: 생존여부 (1: 생존, 0: 사망)
- pclass: 좌석 등급 (1등급, 2등급, 3등급)
- sex: 성별
- age: 나이
- sibsp: 형제 + 배우자 수
- parch: 부모 + 자녀 수
- fare: 좌석 요금
- embarked: 탑승 항구 (S, C, Q)
- class: pclass와 동일
- who: 남자(man), 여자(woman), 아이(child)
- adult_male: 성인 남자 여부
- deck: 데크 번호 (알파벳 + 숫자 혼용)
- embark_town: 탑승 항구 이름
- alive: 생존여부 (yes, no)
- alone: 혼자 탑승 여부

## apply() - 함수를 적용

`apply()`는 데이터 전처리시 굉장히 많이 활용하는 기능입니다.

좀 더 복잡한 **logic을 컬럼 혹은 DataFrame에 적용**하고자 할 때 사용합니다.

In [3]:
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

**who** 컬럼에 대하여 man은 남자, woman은 여자, child는 아이로 변경하고자 한다면 apply를 활용하여 해결할 수 있습니다.

In [4]:
df['who'].value_counts()

man      537
woman    271
child     83
Name: who, dtype: int64

### 함수(function) 정의

In [5]:
# 함수의 인자는 반드시 한 개!
def transform_who(x):
    if x == 'man':
        return '남자'
    elif x == 'woman':
        return '여자'
    else:
        return '아이'

In [6]:
df['who'].apply(transform_who)

0      남자
1      여자
2      여자
3      여자
4      남자
       ..
886    남자
887    여자
888    여자
889    남자
890    남자
Name: who, Length: 891, dtype: object

분포를 확인하면 다음과 같습니다.

In [7]:
df['who'].apply(transform_who).value_counts()

남자    537
여자    271
아이     83
Name: who, dtype: int64

In [9]:
def transform_who(x):
    return x['fare'] / x['age']

In [ ]:
df.apply(transform_who, axis=1)

## apply() - lambda 함수

간단한 logic은 함수를 굳이 정의하지 않고, lambda 함수로 쉽게 해결할 수 있습니다.

In [ ]:
df['survived'].value_counts()

**0: 사망, 1: 생존** 으로 변경하도록 하겠습니다.

In [ ]:
df.head()

In [ ]:
df['survived'].apply(lambda x: '생존' if x == 1 else '사망')

In [ ]:
df['survived'].apply(lambda x: '생존' if x == 1 else '사망').value_counts()

### 연습문제

In [10]:
sample = df.copy()
sample.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [11]:
sample['class'].value_counts()

Third     491
First     216
Second    184
Name: class, dtype: int64

`sample` 데이터프레임에 대하여 `apply()`를 활용하여 `class` 컬럼의 값을 다음과 같이 바꾸고, 분포를 출력후 변경 전과 동일한지 확인하세요

In [13]:
# 코드를 입력해 주세요
def transform_class(x):
    if x == "First":
        return "일등석"
    elif x == "Second":
        return "이등석"
    else:
        return "삼등석"

In [15]:
sample['class'].apply(transform_class).value_counts()

삼등석    491
일등석    216
이등석    184
Name: class, dtype: int64

<p><strong>[출력 결과]</strong></p><pre>삼등석    491
일등석    216
이등석    184
Name: class, dtype: int64</pre>

## groupby() - 그룹

데이터를 특정 기준으로 그룹핑할 때 활용합니다. 엑셀의 피봇테이블과 유사합니다.

In [ ]:
df.head()

타이타닉 호의 생존자와 사망자를 **성별** 기준으로 그룹핑하여 **평균**을 살펴보겠습니다.

In [ ]:
df.groupby('sex').mean()

`groupby()`를 사용할 때는 반드시 aggregate 하는 **통계함수와 일반적으로 같이 적용**합니다.

### 2개 이상의 컬럼으로 그룹

2개 이상의 컬럼으로 그룹핑할 때도 list로 묶어서 지정하면 됩니다.

In [16]:
# 성별, 좌석등급 별 통계
df.groupby(['sex', 'pclass']).mean()

survived        age     sibsp     parch        fare  \
sex    pclass                                                        
female 1       0.968085  34.611765  0.553191  0.457447  106.125798   
       2       0.921053  28.722973  0.486842  0.605263   21.970121   
       3       0.500000  21.750000  0.895833  0.798611   16.118810   
male   1       0.368852  41.281386  0.311475  0.278689   67.226127   
       2       0.157407  30.740707  0.342593  0.222222   19.741782   
       3       0.135447  26.507589  0.498559  0.224784   12.661633   

               adult_male     alone  
sex    pclass                        
female 1         0.000000  0.361702  
       2         0.000000  0.421053  
       3         0.000000  0.416667  
male   1         0.975410  0.614754  
       2         0.916667  0.666667  
       3         0.919308  0.760807

### 1개의 특정 컬럼에 대한 결과 도출

우리의 주요 관심사는 `survived` 컬럼입니다. 만약 `survived`컬럼에 대한 결과만 도출하고 싶다면 컬럼을 맨 끝에 지정합니다.

In [25]:
# 성별, 좌석등급 별 통계
df.groupby(['sex', 'pclass'])['survived'].mean()

sex     pclass
female  1         0.968085
        2         0.921053
        3         0.500000
male    1         0.368852
        2         0.157407
        3         0.135447
Name: survived, dtype: float64

예쁘게 출력하려면 `pd.DataFrame()`으로 감싸주거나, `survived` 컬럼을 []로 한 번 더 감싸주면 됩니다.

In [ ]:
# 성별, 좌석등급 별 통계
df.groupby(['sex', 'pclass'])['survived'].mean()

In [ ]:
# DataFrame으로 출력
pd.DataFrame(df.groupby(['sex', 'pclass'])['survived'].mean())

In [19]:
# DataFrame으로 출력
df.groupby(['sex', 'pclass'])[['survived']].mean()

survived
sex    pclass          
female 1       0.968085
       2       0.921053
       3       0.500000
male   1       0.368852
       2       0.157407
       3       0.135447

### reset_index(): 인덱스 초기화

`reset_index()`: 그룹핑된 데이터프레임의 **index를 초기화**하여 새로운 데이터프레임을 생성합니다.

In [20]:
# index 초기화
df.groupby(['sex', 'pclass'])['survived'].mean().reset_index(drop=True)

0    0.968085
1    0.921053
2    0.500000
3    0.368852
4    0.157407
5    0.135447
Name: survived, dtype: float64

### 다중 컬럼에 대한 결과 도출

끝에 단일 컬럼이 아닌 여러 개의 컬럼을 지정합니다.

In [27]:
# 성별, 좌석등급 별 통계
df.groupby(['sex', 'pclass'])[['survived', 'age']].mean()

survived        age
sex    pclass                     
female 1       0.968085  34.611765
       2       0.921053  28.722973
       3       0.500000  21.750000
male   1       0.368852  41.281386
       2       0.157407  30.740707
       3       0.135447  26.507589

### 다중 통계 함수 적용

여러 가지의 통계 값을 적용할 때는 `agg()`를 사용합니다.

In [ ]:
# 성별, 좌석등급 별 통계
df.groupby(['sex', 'pclass'])[['survived', 'age']].agg(['mean', 'sum'])

## 연습문제

In [32]:
sample = df.copy()
sample

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

`groupby()`를 활용하여 다음을 출력 하세요

`pclass` 별 생존율

In [35]:
# 코드를 입력해 주세요df.groupby('sex').mean()
sample.groupby('pclass')['survived'].mean()

pclass
1    0.629630
2    0.472826
3    0.242363
Name: survived, dtype: float64

<p><strong>[출력 결과]</strong></p><pre>pclass
1    0.629630
2    0.472826
3    0.242363
Name: survived, dtype: float64</pre>

`embarked` 별 생존율 통합 통계

In [38]:
# 코드를 입력해 주세요
sample.groupby('embarked')["survived"].agg(["mean","var"])

mean       var
embarked                    
C         0.553571  0.248610
Q         0.389610  0.240943
S         0.336957  0.223764

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mean</th>
      <th>var</th>
    </tr>
    <tr>
      <th>embarked</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>C</th>
      <td>0.553571</td>
      <td>0.248610</td>
    </tr>
    <tr>
      <th>Q</th>
      <td>0.389610</td>
      <td>0.240943</td>
    </tr>
    <tr>
      <th>S</th>
      <td>0.336957</td>
      <td>0.223764</td>
    </tr>
  </tbody>
</table>
</div>

`who`, `pclass`별 생존율, 생존자수

In [39]:
# 코드를 입력해 주세요
sample.groupby(['who','pclass'])["survived"].agg(["mean","sum"])

mean  sum
who   pclass               
child 1       0.833333    5
      2       1.000000   19
      3       0.431034   25
man   1       0.352941   42
      2       0.080808    8
      3       0.119122   38
woman 1       0.978022   89
      2       0.909091   60
      3       0.491228   56

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>mean</th>
      <th>sum</th>
    </tr>
    <tr>
      <th>who</th>
      <th>pclass</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="3" valign="top">child</th>
      <th>1</th>
      <td>0.833333</td>
      <td>5</td>
    </tr>
    <tr>
      <th>2</th>
      <td>1.000000</td>
      <td>19</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.431034</td>
      <td>25</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">man</th>
      <th>1</th>
      <td>0.352941</td>
      <td>42</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.080808</td>
      <td>8</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.119122</td>
      <td>38</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">woman</th>
      <th>1</th>
      <td>0.978022</td>
      <td>89</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.909091</td>
      <td>60</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.491228</td>
      <td>56</td>
    </tr>
  </tbody>
</table>
</div>

- 남자의 나이는 남자 나이의 평균으로 채우세요
- 여자의 나이는 여자 나이의 평균으로 채우세요

In [40]:
# 결측치 확인
print(sample['age'].isnull().sum())
print(f"age 평균: {sample['age'].mean():.2f}")

177
age 평균: 29.70


In [ ]:
# 코드를 입력해 주세요
# 결측치 평균으로 채우기! 중요!
sample.groupby('sex')['age'].apply(lambda x:x.fillna(x.mean()))

In [ ]:
# 검증코드
print(sample['age'].isnull().sum())
print(f"age 평균: {sample['age'].mean():.2f}")

<p><strong>[출력 결과]</strong></p><pre>0
age 평균: 29.74
</pre>

## pivot_table()

피벗테이블은 엑셀의 피벗과 동작이 유사하며, `groupby()`와도 동작이 유사합니다.

기본 동작 원리는 `index`, `columns`, `values`를 지정하여 피벗합니다.

### 1개 그룹에 대한 단일 컬럼 결과

In [41]:
# index에 그룹을 표기
# index : 행, Colummns : 열, Values : 값
df.pivot_table(index='who', values='survived')

survived
who            
child  0.590361
man    0.163873
woman  0.756458

In [42]:
# columns에 그룹을 표기
df.pivot_table(columns='who', values='survived')

who          child       man     woman
survived  0.590361  0.163873  0.756458

### 다중 그룹에 대한 단일 컬럼 결과

In [43]:
df.pivot_table(index=['who', 'pclass'], values='survived')

survived
who   pclass          
child 1       0.833333
      2       1.000000
      3       0.431034
man   1       0.352941
      2       0.080808
      3       0.119122
woman 1       0.978022
      2       0.909091
      3       0.491228

### index에 컬럼을 중첩하지 않고 행과 열로 펼친 결과

In [ ]:
df.pivot_table(index='who', columns='pclass', values='survived')

### 다중 통계함수 적용

In [ ]:
df.pivot_table(index='who', columns='pclass', values='survived', aggfunc=['sum', 'mean'])

### 연습문제

In [44]:
tips = sns.load_dataset('tips')
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

`tips` 데이터셋을 활용하여 다음을 출력하세요

다음의 pivot table을 생성합니다.
- value는 `tip`에 대한 평균값을 산출합니다.

In [48]:
# 코드를 입력해 주세요
tips.pivot_table(index='smoker',columns='day',values='tip')

day         Thur     Fri       Sat       Sun
smoker                                      
Yes     3.030000  2.7140  2.875476  3.516842
No      2.673778  2.8125  3.102889  3.167895

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>day</th>
      <th>Thur</th>
      <th>Fri</th>
      <th>Sat</th>
      <th>Sun</th>
    </tr>
    <tr>
      <th>smoker</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Yes</th>
      <td>3.030000</td>
      <td>2.7140</td>
      <td>2.875476</td>
      <td>3.516842</td>
    </tr>
    <tr>
      <th>No</th>
      <td>2.673778</td>
      <td>2.8125</td>
      <td>3.102889</td>
      <td>3.167895</td>
    </tr>
  </tbody>
</table>
</div>

다음의 pivot table을 생성합니다.
- value는 `total_bill`에 대한 평균과 합계를 산출합니다.

In [52]:
# 코드를 입력해 주세요
a = tips.pivot_table(index='day', columns='time', values='total_bill', aggfunc=['mean', 'sum'])
a.loc[:, ('sum', 'Lunch')] = a.loc[:, ('sum', 'Lunch')].fillna(0)
a

mean                 sum         
time      Lunch     Dinner    Lunch   Dinner
day                                         
Thur  17.664754  18.780000  1077.55    18.78
Fri   12.845714  19.663333    89.92   235.96
Sat         NaN  20.441379      NaN  1778.40
Sun         NaN  21.410000      NaN  1627.16

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead tr th {
        text-align: left;
    }

    .dataframe thead tr:last-of-type th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr>
      <th></th>
      <th colspan="2" halign="left">mean</th>
      <th colspan="2" halign="left">sum</th>
    </tr>
    <tr>
      <th>time</th>
      <th>Lunch</th>
      <th>Dinner</th>
      <th>Lunch</th>
      <th>Dinner</th>
    </tr>
    <tr>
      <th>day</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Thur</th>
      <td>17.664754</td>
      <td>18.780000</td>
      <td>1077.55</td>
      <td>18.78</td>
    </tr>
    <tr>
      <th>Fri</th>
      <td>12.845714</td>
      <td>19.663333</td>
      <td>89.92</td>
      <td>235.96</td>
    </tr>
    <tr>
      <th>Sat</th>
      <td>NaN</td>
      <td>20.441379</td>
      <td>0.00</td>
      <td>1778.40</td>
    </tr>
    <tr>
      <th>Sun</th>
      <td>NaN</td>
      <td>21.410000</td>
      <td>0.00</td>
      <td>1627.16</td>
    </tr>
  </tbody>
</table>
</div>